feat: add tibia_urls field into information-section

Kai-Animator opened the pull request:
This pull request addresses **Issue #381** by adding the `Link` field to the `Information` section in all API responses. The `Link` field provides a direct URL to the source data on the official Tibia website.

**Changes Overview:**

- **Updated `Information` Struct:**
  - Added a new `Link` field to the `Information` struct in `webserver.go`:
    ```go
    type Information struct {
        APIDetails APIDetails `json:"api"`       // The API details.
        Timestamp  string     `json:"timestamp"` // The timestamp from when the data was processed.
        Link       string     `json:"link"`      // The link to the source of the data.
        Status     Status     `json:"status"`    // The response status information.
    }
    ```
- **Modified API Handlers:**
  - Updated all API endpoint handlers to populate the `Link` field with the appropriate URL pointing to the source data.
  
  - Example from `TibiaWorldsWorld.go`:
  
    ```go
    return WorldResponse{
        World: World{
            // ... existing fields ...
        },
        Information: Information{
            APIDetails: TibiaDataAPIDetails,
            Timestamp:  TibiaDataDatetime(""),
            Link:       "https://www.tibia.com/community/?subtopic=worlds&world=" + world,
            Status: Status{
                HTTPCode: http.StatusOK,
            },
        },
    }, nil
    ```

**Testing:**

- Tested all modified endpoints to ensure the `Link` field is correctly populated.
- Verified that the URLs are accurate and lead to the correct pages on Tibia.com.

---

**Notes:**

- NewsList was kind of tricky to handle the different patterns so I decided to sum all different URLs into only "https://www.tibia.com/news/?subtopic=newsarchive"

- This is my first PR in the project so I am open to feedback and willing to make any necessary changes to meet the project's standards.

fix #381

tobiasehlert commented on src/webserver.go:
```diff
@@ -47,6 +47,7 @@ type OutInformation struct {
 type Information struct {
 	APIDetails APIDetails `json:"api"`       // The API details.
 	Timestamp  string     `json:"timestamp"` // The timestamp from when the data was processed.
+	Link       string     `json:"link"`      // The link to the source of the data.
```
```suggestion
	TibiaURL   []string   `json:"tibia_url"` // The links to the sources of the data on tibia.com.
```

tobiasehlert commented on src/webserver.go:
```diff
@@ -121,6 +122,7 @@ func runWebServer() {
 		data := Information{
 			APIDetails: TibiaDataAPIDetails,
 			Timestamp:  TibiaDataDatetime(""),
+			Link:       "-",
```
```suggestion
			TibiaURL:   []string{},
```

tobiasehlert commented on src/webserver.go:
```diff
@@ -1089,6 +1091,7 @@ func TibiaDataErrorHandler(c *gin.Context, err error, httpCode int) {
 	info := Information{
 		APIDetails: TibiaDataAPIDetails,
 		Timestamp:  TibiaDataDatetime(""),
+		Link:       "-",
```
```suggestion
		TibiaURL:   []string{},
```

tobiasehlert reviewed (changes requested):
Thanks for the PR @Kai-Animator, this is a good start!
There are some things that I'd like to be adjusted.

One of them is that there is already a link for every endpoints source available, so hardcoding the line in the "Link" field is not optimal. It would be better to have it being populated by the value set in "tibiadataRequest.URL" before being sent to tibiaDataRequestHandler.

For instance here:
https://github.com/TibiaData/tibiadata-api-go/blob/a4afbe1bf78013effcdf6e71605d31811374531b/src/webserver.go#L394-L407

I'd also like the "Link" to be named "TibiaURL" instead, since that's a slightly better description of that it is.
Also, we should probably want to have a []string as well, in case we add multi-upstream sources.

Kai-Animator commented:
Thanks for the review @tobiasehlert . I will work on it 👍

Kai-Animator commented:
Hi @tobiasehlert  :wave: 

I’ve implemented the changes you requested:

- The "Link" field is now populated using the value from `tibiadataRequest.URL` before it's sent to `tibiaDataRequestHandler`, rather than hardcoding the URL. (Newslist and HousesOverview might need refactor)
- Renamed "Link" to "TibiaURL".
- I've also updated the field to support a `[]string` for multi-upstream sources in the future (HousesOverview is using multiple URLs now).

Please let me know if there’s anything else :]

phenpessoa commented:
Since this is now slice, I think it should probably be named `TibiaURLs`, plural.

tobiasehlert commented:
Yeah good input @phenpessoa. 👍

phenpessoa commented:
The field name was updated but the json is still called `tibia_url`. I think it should probably be `tibia_urls`.

Kai-Animator commented:
> The field name was updated but the json is still called `tibia_url`. I think it should probably be `tibia_urls`.

Hey @phenpessoa, thanks for looking at the code, is this the change that you mentioned?

phenpessoa commented:
Yup! Looks good.

tobiasehlert commented on src/TibiaNews.go:
```diff
@@ -16,7 +16,7 @@ type News struct {
 	Category    string `json:"category"`        // The category of the news.
 	Type        string `json:"type,omitempty"`  // The type of news.
-	TibiaURL    string `json:"url"`             // The URL for the news with id.
+	TibiaURLs    string `json:"url"`             // The URL for the news with id.
```
This is supposed to be one URL for the news entry.

tobiasehlert commented on src/TibiaNews.go:
```diff
@@ -45,13 +45,12 @@ func TibiaNewsImpl(NewsID int, rawUrl string, BoxContentHTML string) (NewsRespon
 
 	NewsData.ID = NewsID
-	NewsData.TibiaURL = rawUrl
+	NewsData.TibiaURLs = rawUrl
```
This is supposed to be one here as well.

tobiasehlert commented on src/TibiaNews_test.go:
```diff
@@ -61,7 +64,7 @@ func TestNews6512(t *testing.T) {
 	assert.Empty(newsArticleJson.News.Title)
 	assert.Equal("community", newsArticleJson.News.Category)
 	assert.Equal("ticker", newsArticleJson.News.Type)
-	assert.Equal("https://www.tibia.com/news/?subtopic=newsarchive&id=6512", newsArticleJson.News.TibiaURL)
+	assert.Equal("https://www.tibia.com/news/?subtopic=newsarchive&id=6512", newsArticleJson.News.TibiaURLs)
```
Same as above

tobiasehlert commented on src/TibiaNews_test.go:
```diff
@@ -26,13 +26,16 @@ func TestNewsById(t *testing.T) {
 	assert.Equal(6529, newsArticleJson.News.ID)
 	assert.Equal("2022-01-12", newsArticleJson.News.Date)
 	assert.Empty(newsArticleJson.News.Title)
 	assert.Equal("development", newsArticleJson.News.Category)
 	assert.Equal("ticker", newsArticleJson.News.Type)
-	assert.Equal("https://www.tibia.com/news/?subtopic=newsarchive&id=6529", newsArticleJson.News.TibiaURL)
+	assert.Equal("https://www.tibia.com/news/?subtopic=newsarchive&id=6529", newsArticleJson.News.TibiaURLs)
```
Same as above

tobiasehlert commented on src/TibiaNews_test.go:
```diff
@@ -90,7 +93,7 @@ func TestNews504(t *testing.T) {
 	assert.Empty(newsArticleJson.News.Title)
 	assert.Equal("community", newsArticleJson.News.Category)
 	assert.Equal("ticker", newsArticleJson.News.Type)
-	assert.Equal("https://www.tibia.com/news/?subtopic=newsarchive&id=504", newsArticleJson.News.TibiaURL)
+	assert.Equal("https://www.tibia.com/news/?subtopic=newsarchive&id=504", newsArticleJson.News.TibiaURLs)
```
Same as above

tobiasehlert commented on src/TibiaNews_test.go:
```diff
@@ -119,7 +122,7 @@ func TestNews6481(t *testing.T) {
 	assert.Equal("New Mounts", newsArticleJson.News.Title)
 	assert.Equal("development", newsArticleJson.News.Category)
 	assert.Empty(newsArticleJson.News.Type)
-	assert.Equal("https://www.tibia.com/news/?subtopic=newsarchive&id=6481", newsArticleJson.News.TibiaURL)
+	assert.Equal("https://www.tibia.com/news/?subtopic=newsarchive&id=6481", newsArticleJson.News.TibiaURLs)
```
Same as above

tobiasehlert commented on src/TibiaNewslist.go:
```diff
@@ -16,7 +16,7 @@ type NewsItem struct {
 	Category string `json:"category"`          // The category of the news.
 	Type     string `json:"type"`              // The type of news.
-	TibiaURL string `json:"url"`               // The URL for the news with id.
+	TibiaURLs string `json:"url"`               // The URL for the news with id.
```
Same as above

tobiasehlert commented on src/TibiaNewslist.go:
```diff
@@ -63,7 +63,7 @@ func TibiaNewslistImpl(days int, BoxContentHTML string) (NewsListResponse, error
 		NewsSplit := strings.Split(NewsURL, NewsID)
 		OneNews.ID = TibiaDataStringToInteger(NewsID)
-		OneNews.TibiaURL = NewsSplit[0] + NewsID
+		OneNews.TibiaURLs = NewsSplit[0] + NewsID
```
Same as above

tobiasehlert commented on src/TibiaNewslist_test.go:
```diff
@@ -37,6 +40,6 @@ func TestNewsList(t *testing.T) {
 	assert.Equal("A number of issues related to the 25 years activities have been fixed,...", firstArticle.News)
 	assert.Equal("development", firstArticle.Category)
 	assert.Equal("ticker", firstArticle.Type)
-	assert.Equal("https://www.tibia.com/news/?subtopic=newsarchive&id=6529", firstArticle.TibiaURL)
+	assert.Equal("https://www.tibia.com/news/?subtopic=newsarchive&id=6529", firstArticle.TibiaURLs)
```
Same as above

tobiasehlert reviewed (commented):
We are getting closer! 🚀
The news endpoints just need some adjustment, but then we should be really to rock.

Kai-Animator commented:
Thanks for the feedback @tobiasehlert, sorry search and replace all because I though you weren't using TibiaURL in other files. 🙇 Hope this revisions are satisfying.

tobiasehlert reviewed (approved):
looks good to me 🌟
